Implement block range checksum validation

jhermsmeier opened the pull request:
Connects to #6 

This adds block range checksum calculation and verification to `BlockMap.ReadStream()` and `BlockMap.FilterStream()`. Upon encountering a read / write of a range with a checksum mismatch an `error` event will be emitted, where the error will not end the stream, and if handled via an event handler can occur multiple times.

With this we can detect corrupt source images, or verify an already flashed image by simply reading it back through a `BlockMap.ReadStream()` with verification enabled.

Verification is enabled by default, but can be disabled by passing `{ verify: false }` in the options.

jviotti commented on lib/filter-stream.js:
```diff
@@ -69,22 +95,51 @@ function FilterStream( blockMap, options ) {
    * @return {Boolean}
    */
-  _pushBlock( block ) {
+  _pushBlock( block, next ) {
```
Is `next` used anywhere?

jhermsmeier replied:
Nope, removed it, that was a leftover from a previous iteration 😅

jviotti commented:
Looking good. Some testing ideas:

- Can we test that if `verify` is `false`, no error will be raised?
- Can we test that if we omit the options, `verify` is `true` by default?
- Can we test this for `sha1` as well (used in older bmap versions)?

jhermsmeier commented:
> Can we test that if verify is false, no error will be raised?
> Can we test this for sha1 as well (used in older bmap versions)?

Added tests for both cases by adding tests across all versions.

> Can we test that if we omit the options, verify is true by default?

The options are omitted in the invalid checksum detection tests, so that is tested.
I added `{ verify: true }` to the second one, so that that case is covered as well now.

lurch commented:
Doesn't really matter (and the answer may well be "that wouldn't be worth it") but I wonder if this PR could have been implemented with more code-sharing between filter-stream and read-stream?

jhermsmeier commented:
I did consider doing that at first, but the overlap is actually pretty small, and the logic is quite different in this one (due to order and dependency of events); it didn't pay off (would've introduced more complexity than reducing code).
